crates/handlers/src/upstream_oauth2/callback.rs

```diff
@@ -212,32 +214,54 @@ pub(crate) async fn get(
         redirect_uri,
     };
 
-    let id_token_verification_data = JwtVerificationData {
+    let verification_data = JwtVerificationData {
         issuer: &provider.issuer,
         jwks: &jwks,
         // TODO: make that configurable
         signing_algorithm: &mas_iana::jose::JsonWebSignatureAlg::Rs256,
         client_id: &provider.client_id,
     };
 
-    let (response, id_token) =
+    let (response, id_token_map) =
         mas_oidc_client::requests::authorization_code::access_token_with_authorization_code(
             &http_service,
             client_credentials,
             lazy_metadata.token_endpoint().await?,
             code,
             validation_data,
-            Some(id_token_verification_data),
+            Some(verification_data),
             clock.now(),
             &mut rng,
         )
         .await?;
 
-    let (_header, id_token) = id_token.ok_or(RouteError::MissingIDToken)?.into_parts();
+    let (_header, id_token) = id_token_map
+        .clone()
+        .ok_or(RouteError::MissingIDToken)?
+        .into_parts();
+
+    let use_userinfo_endpoint = match provider.user_profile_method {
+        UpstreamOAuthProviderUserProfileMethod::Auto => !provider.scope.contains("openid"),
+        UpstreamOAuthProviderUserProfileMethod::UserinfoEndpoint => true,
+    };
+
+    let userinfo = if use_userinfo_endpoint {
+        let user_info_resp = mas_oidc_client::requests::userinfo::fetch_userinfo(
+            &http_service,
+            lazy_metadata.userinfo_endpoint().await?,
+            response.access_token.as_str(),
+            Some(verification_data),
+            &id_token_map.ok_or(RouteError::MissingIDToken)?,
+        )
+        .await?;
+        minijinja::Value::from_serialize(&user_info_resp)
+    } else {
+        minijinja::Value::from_serialize(&id_token)
+    };
 
     let env = {
         let mut env = environment();
-        env.add_global("user", minijinja::Value::from_serialize(&id_token));
+        env.add_global("user", userinfo);
```

Comment:
Just thinking out loud at this point, but it would be nice to have both the userinfo and the idtoken claims in the environment.

I don't know what the right config is to tell whether to fetch userinfo or not for example, *but* I would set `userinfo_claims` in the template environment if we fetched it, `id_token_claims` if we have an `id_token` and `user` being somehow a merge of the two?

One stretch goal of this would be for this to work with non-OIDC OAuth 2.0 providers, like GitHub, so we'd need to remove the id_token requirement at some point as well